### DISCUSSION TITLE
ControlledTree load Data

### DISCUSSION THREAD

**Discussion by PaoloPetra** [iTwin/itwinjs-core#3095] [Q&A]
Goodmorning everyone
I am trying to visualize data inside a tree, I am looking at the examples on the site where everything works correctly but in my program I can't load fake sample data.

I recall this class
const dataProvider = React.useMemo (() => new NodesWithCheckboxProvider (), []);
which goes to implement the various nodes as in your very detailed examples
but from this class it seems that nothing uploaded comes back as attached images

thanks for your invaluable help.

I am attaching two images of the console.

![1](https://user-images.githubusercontent.com/40106886/150966967-981d5b84-6ba8-4b25-8921-35257ee8598a.PNG)
![2](https://user-images.githubusercontent.com/40106886/150966986-a040b8c0-4234-48b1-b8d8-e0dabc39c2d6.PNG)

**Comment by grigasp** [comment]
Any chance you could create [a sandbox](https://www.itwinjs.org/sandboxes/) to demonstrate the issue? Or at least share the code of your component?

**Comment by PaoloPetra** [comment]
hello of course for convenience I put everything inside a single file the component <PropertyTreeComponent> is called inside a modal. here is the code 👍 

import React, { FunctionComponent } from "react";
import { ControlledTree, SelectionMode, useTreeEventsHandler, useTreeModelSource, useTreeNodeLoader, useVisibleTreeNodes } from "@bentley/ui-components";
import { DelayLoadedTreeNodeItem, ITreeDataProvider, TreeDataChangesListener, TreeNodeItem } from "@bentley/ui-components";
import { PropertyRecord } from "@bentley/ui-abstract";
import { BeEvent } from "@bentley/bentleyjs-core";

export const PropertyTreeComponent: FunctionComponent = () => {
  // create data provider to get some nodes to show in tree
  // `React.useMemo' is used avoid creating new object on each render
  const dataProvider = React.useMemo(() => new NodesWithCheckboxProvider(), []);
  const [treeDataProvider, setTreeDataProvider] = React.useState(new SampleDataProvider());
  // create model source for tree. Model source contains tree model and provides an API to modify
  // model and listen for changes in tree model.
  // `useTreeModelSource` creates new model source when 'dataProvider' object changes
  const modelSource = useTreeModelSource(treeDataProvider);

  // create tree node loader to load nodes to tree model. It uses 'dataProvider' to get
  // nodes and adds them to tree model using 'modelSource'
  const nodeLoader = useTreeNodeLoader(treeDataProvider, modelSource);

  // create parameters for default tree event handler. It needs 'nodeLoader' to load child nodes when parent node
  // is expanded and 'modelSource' to modify nodes' state in tree model. 'collapsedChildrenDisposalEnabled' flag
  // specifies that child nodes should be removed from tree model when parent node is collapsed.
  // `React.useMemo' is used to avoid creating new params object on each render
  const eventHandlerParams = React.useMemo(() => ({ nodeLoader, modelSource, collapsedChildrenDisposalEnabled: true }), [nodeLoader, modelSource]);

  // create default event handler. It handles tree events: expanding/collapsing, selecting/deselecting nodes,
  // checking/unchecking checkboxes. It also initiates child nodes loading when parent node is expanded.
  // `useTreeEventsHandler` created new event handler when 'eventHandlerParams' object changes
  const eventHandler = useTreeEventsHandler(eventHandlerParams);

  // get list of visible nodes to render in `ControlledTree`. This is a flat list of nodes in tree model.
  // `useVisibleTreeNodes` uses 'modelSource' to get flat list of nodes and listens for model changes to
  // re-render component with updated nodes list
  const visibleNodes = useVisibleTreeNodes(modelSource);

  return <>
    <div className="row"  >
        <div className="col-10 col.delete">
            <ControlledTree
                nodeLoader={nodeLoader}
                selectionMode={SelectionMode.None}
                treeEvents={eventHandler}
                visibleNodes={visibleNodes}
            />
        </div>
    </div>
  </>;
};


export class SampleDataProvider implements ITreeDataProvider {
    public onTreeNodeChanged = new BeEvent<TreeDataChangesListener>();
  
    public async getNodesCount(parent?: TreeNodeItem) {
      if (parent === undefined)
        return 5;
  
      switch (parent.id) {
        case "TestNode-1": return 3;
        case "TestNode-2": return 3;
        case "TestNode-2-2": return 2;
        case "TestNode-2-3": return 2;
        case "TestNode-3": return 3;
        case "TestNode-5": return 1;
        default: return 0;
      }
    }
  
    public async getNodes(parent?: TreeNodeItem) {
      if (parent === undefined) {
        return [
          createNode("TestNode-1", "TestNode 1", true), createNode("TestNode-2", "TestNode 2", true), createNode("TestNode-3", "TestNode 3", true),
          createNode("TestNode-4", "TestNode 4"), createNode("TestNode-5", "TestNode 5", true),
        ];
      }
  
      switch (parent.id) {
        case "TestNode-1": return [createNode("TestNode-1-1", "TestNode 1-1"), createNode("TestNode-1-2", "TestNode 1-2"), createNode("TestNode-1-3", "TestNode 1-3")];
        case "TestNode-2": return [createNode("TestNode-2-1", "TestNode 2-1"), createNode("TestNode-2-2", "TestNode 2-2", true), createNode("TestNode-2-3", "TestNode 2-3", true)];
        case "TestNode-2-2": return [createNode("TestNode-2-2-1", "TestNode 2-2-1"), createNode("TestNode-2-2-2", "TestNode 2-2-2")];
        case "TestNode-2-3": return [createNode("TestNode-2-3-1", "TestNode 2-3-1"), createNode("TestNode-2-3-2", "TestNode 2-3-2")];
        case "TestNode-3": return [createNode("TestNode-3-1", "TestNode 3-1"), createNode("TestNode-3-2", "TestNode 3-2"), createNode("TestNode-3-3", "TestNode 3-3")];
        case "TestNode-5": return [createNode("TestNode-5-1", "TestNode 5-1")];
        default: return [];
      }
    }
  }
  
  const createNode = (id: string, label: string, hasChildren?: boolean): DelayLoadedTreeNodeItem => {
    return {
      id,
      label: PropertyRecord.fromString(label),
      hasChildren,
    };
  };

  class NodesWithCheckboxProvider extends SampleDataProvider {
    public async getNodes(parent?: TreeNodeItem) {
      // get nodes from SampleDataProvider
      const nodes = await super.getNodes(parent);
  
      // for each node set 'isCheckboxVisible' flag to true
      nodes.forEach((node) => {
        node.isCheckboxVisible = true;
      });
      return nodes;
    }
  }

**Reply by PaoloPetra** [reply]
sorry there is an error in the file I use dataProvider and not treeDataProvider but the result does not change

**Comment by grigasp** [comment]
As far as I can tell the problem is that the tree container has zero size. The `ControlledTree` component takes full width and height of its parent container, so it will not render anything if there's no size given to it. 

Have a look at this: https://www.itwinjs.org/sandboxes/grigas/Catlaina-the-pink-shrimp-406. I assigned width and height to the parent `div` at line 44 there.

**Reply by PaoloPetra** [reply]
oh god what a fool I am
